Split DataContainer to a HDF and non HDF part

pmrv opened the pull request:
Basically DataContainer is renamed to DataContainerBase and the methods used for HasHDF/HDFStub are moved to a new sub class DataContainer. Docstrings are replicated where necessary.  No API changes for users.

In preparation of https://github.com/orgs/pyiron/projects/7/views/1

@liamhuber Pulling `DataContainer` out should be just a matter of copy/paste now.

pmrv committed:
Split DataContainer to a HDF and non HDF part

Basically DataContainer is renamed to DataContainerBase and the methods
used for HasHDF/HDFStub are moved to a new sub class DataContainer.
Docstrings are replicated where necessary.

pyiron-runner committed:
Format black

liamhuber commented:
> @liamhuber Pulling DataContainer out should be just a matter of copy/paste now.

I think this may be of even more imminent utility to @samwaseda?

Will review shortly, but I peeked at the diff and it was what I intuitively expected so that should go pretty smoothly.

liamhuber commented on pyiron_base/storage/datacontainer.py:
```diff
@@ -47,16 +47,16 @@ def _normalize(key):
 
 
-class DataContainer(MutableMapping, Lockable, HasGroups, HasHDF):
+class DataContainerBase(MutableMapping, Lockable, HasGroups):
```
If we move this part off to its own repo (e.g. `pyiron/snippets`) and leave the child class here, then in the other location I would simply call this `DataContainer` and here do like `from pyiron.snippets import DataContainer as DataContainerBase`. In the context of making the break-apart go as smoothly as possible, I am totally behind the naming scheme used in this PR, I just want to double check that we have the same middle-term vision on naming here.

pmrv replied:
Yeah, that's exactly how I thought it would go down.

liamhuber commented on pyiron_base/storage/datacontainer.py:
```diff
@@ -451,7 +430,7 @@ def to_builtin(self, stringify=False):
                 # transparent for the rest of the module
                 k = str(k)
-                if isinstance(v, DataContainer):
+                if isinstance(v, type(self)):
```
Is this (and the other nearby `type(self)` appearances) really correct? I naively would have thought this should be `isinstance(v, DataContainerBase))`, since that is already what guarantees us to have `.to_builtin`. 

What _should_ the behaviour be if a `DataContainer` instance has a `DataContainerBase` instance as an item? I would want it to "just work", but I know that is vague. Here I think it certainly wouldn't as the type check would evaluate `False` and we'd get kicked over to the `repr` branch.

I don't know what the exact right behaviour is, but I think this needs a test added to solidify whatever choice you make.

pmrv replied:
Agree with your reasoning, it seems a more useful choice, I just hadn't considered it.  I've changed it back, but tests will have to come later.

liamhuber commented on pyiron_base/storage/datacontainer.py:
```diff
@@ -566,11 +545,11 @@ def _search_parent(self, key, stop_on_first_hit=True):
         for it in self.groups():
             hit = self[it]._search_parent(key, stop_on_first_hit)
-            if isinstance(hit, DataContainer):
+            if isinstance(hit, type(self)):
```
Similar... if `DataClass` hits a `DataClassBase` instance, do we want to stop or not? Because right now it's "not". I could be convinced that's the right behaviour, but intuitively it feels wrong.

liamhuber commented on pyiron_base/storage/datacontainer.py:
```diff
@@ -920,13 +962,27 @@ def _force_load(self, recursive=True):
         # values are loaded from HDF once they are accessed via __getitem__, which is implicitly called by values()
         for v in self.values():
-            if recursive and isinstance(v, DataContainer):
+            if recursive and isinstance(v, type(self)):
```
Similarly, should this be `(v, DataContainer)`? If my `MyDataContainer(DataContainer)` has `DataContainer` instances as items, don't those instances contain enough power to recurse like this?

liamhuber reviewed (changes requested):
99% as expected and no nits or anything.

The only thing I'm not sure of is the consistent use of `type(self)` for a bunch of recursion checks. Most of these seem to me like they should instead hard-code the class they're written in, so that if derived classes hold an instance of their parent class the recursion can continue uninterrupted. I'm not totally sure that's the right behaviour choice though.

At any rate, now that we explicitly have a base and child class, I think it's important to extend the tests to cover the case of child-holds-parent-instance to encode whatever behaviour decision you make.

samwaseda commented on pyiron_base/storage/datacontainer.py:
```diff
@@ -851,62 +941,14 @@ def normalize_key(name):
+        if not isinstance(value, HDFStub):
+            return value
+        else:
+            value = self[key] = value.load()
+            return value
```
Does it have a particular reason that it assigns `value` here?

pmrv replied:
Once we have loaded a stub, we don't want to load it again a second time, so this overwrites the stub with the actual value.  I guess

```python
return self[key] := value.load()
```

would work as well, but is only available in newer python version (not sure which atm).

samwaseda replied:
ok it's a small thing, but my problem is that by returning `value` it might give the impression that it's the copy of the data, while in reality if you change it, it changes the content in `self[key]` as well. So I would simply set `self[key] = value.load()` and `return self[key]`.

pmrv replied:
Won't work because we are inside `__getitem__`, so using `self[key]` will recurse without stopping.

samwaseda replied:
ah ok. Beautiful

pmrv commented on pyiron_base/storage/datacontainer.py:
```diff
@@ -851,62 +941,14 @@ def normalize_key(name):
+        if not isinstance(value, HDFStub):
+            return value
+        else:
+            value = self[key] = value.load()
+            return value
```
```suggestion
            return self[key] := value.load()
```

the "walrus" is already in since 3.8 it turns out, so let's just use it.

pmrv replied:
Ah, bummer it's not allowed in this place.  Will revert.

niklassiemer reviewed (commented):
LGTM once @liamhuber s concerns are matched

pmrv committed:
Hard code instance checks

If we put a sub class of DataContainer(Base) into a base class, we still want to recurse all of it, so
checking against type(self) is not sufficient, since that would abort early.

pmrv commented:
```
 Traceback (most recent call last):
  File "/home/runner/work/pyiron_base/pyiron_base/tests/flex/test_pythonfunctioncontainer.py", line 63, in test_with_executor_wait
    self.assertTrue(job.server.future.done())
AssertionError: False is not true
```

@jan-janssen Sounds related to your recent developments.  Would that be fixed by #1300 or #1301?

jan-janssen commented:
> @jan-janssen Sounds related to your recent developments. Would that be fixed by #1300 or #1301?

The issue was solved by restarting the CI. We have to take a look at this and maybe increase the time out in case the issue continues to come up.

pyiron-runner committed:
Format black

jan-janssen commented:
Once this is merged, we should also release a new `pyiron_base` version.

pmrv commented:
> > @jan-janssen Sounds related to your recent developments. Would that be fixed by #1300 or #1301?
> 
> The issue was solved by restarting the CI. We have to take a look at this and maybe increase the time out in case the issue continues to come up.

I think it makes more sense to just use the timeout directly in the tests, then this can never happen.

jan-janssen commented:
@pmrv As more tests were requested, I changed the status of this pull request to `draft` until the tests are added.

pmrv committed:
Add tests for to_builtin

pmrv committed:
Simply toy sub class

Setting one value is not used in the older tests and would make the newer ones more complicated

pmrv committed:
Add tests for list_nodes/groups

liamhuber reviewed (approved):
Nice, type checking and testing changes lgtm.
